transport-native-io_uring/src/main/c/netty_io_uring_native.c

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+#define _GNU_SOURCE // RTLD_DEFAULT
+#include "io_uring.h"
+#include <dlfcn.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <jni.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "syscall.h"
+#include <errno.h>
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <syscall.h>
+#include <unistd.h>
+
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+static jmethodID ringBufferMethodId = NULL;
+static jmethodID ioUringSubmissionQueueMethodId = NULL;
+static jmethodID ioUringCommpletionQueueMethodId = NULL;
+static jclass ringBufferClass = NULL;
+static jclass ioUringCompletionQueueClass = NULL;
+static jclass ioUringSubmissionQueueClass = NULL;
+
+void io_uring_unmap_rings(struct io_uring_sq *sq, struct io_uring_cq *cq) {
+    munmap(sq->ring_ptr, sq->ring_sz);
+    if (cq->ring_ptr && cq->ring_ptr != sq->ring_ptr) {
+        munmap(cq->ring_ptr, cq->ring_sz);
+  }
+}
+
+static int io_uring_mmap(int fd, struct io_uring_params *p,
+			 struct io_uring_sq *sq, struct io_uring_cq *cq)
+{
+	size_t size;
+	int ret;
+
+	sq->ring_sz = p->sq_off.array + p->sq_entries * sizeof(unsigned);
+	cq->ring_sz = p->cq_off.cqes + p->cq_entries * sizeof(struct io_uring_cqe);
+
+	if (p->features & IORING_FEAT_SINGLE_MMAP) {
```

Comment:
nit: usually we prefer to use `== 1` or `!= 0` in branches. Consider doing so everywhere (there are multiple lines of code that are affected.